{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Align Fasta to Reference Fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current machine: DESKTOP-5RJD9NC\n",
      "D:\\gitProjects\\covid19_data\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import os as os\n",
    "import sys as sys\n",
    "import multiprocessing as mp\n",
    "import pandas as pd\n",
    "import socket\n",
    "import matplotlib.pyplot as plt\n",
    "from itertools import groupby\n",
    "import os\n",
    "import re as re\n",
    "\n",
    "### Pick the right path (whether on cluster or at home)\n",
    "socket_name = socket.gethostname()\n",
    "print(f\"Current machine: {socket_name}\")\n",
    "if socket_name == \"DESKTOP-5RJD9NC\":\n",
    "    path = \"/gitProjects/covid19_data\"   # The Path on Harald's machine\n",
    "else: \n",
    "    raise RuntimeWarning(\"Not compatible machine. Check!!\")\n",
    "\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean Fasta File (remove is)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_fasta(sample_line, seq_lines, savepath=\"\", output=True):\n",
    "    \"\"\"Write a single Fasta file.\n",
    "    sample_line: Line of the Sample to write\n",
    "    seq_lines: Genotype Sequences to write\n",
    "    savepath: Where to write Fasta File to\n",
    "    \"\"\"\n",
    "    f1 = open(savepath, \"w\")\n",
    "    f1.write(sample_line)  # Write the new line\n",
    "    for line in seq_lines:\n",
    "        f1.write(line)\n",
    "    f1.close()\n",
    "    if output:\n",
    "        print(f\"Saved fasta to {savepath}\")\n",
    "    \n",
    "def fasta_iter_raw(fasta_name):\n",
    "    \"\"\"\n",
    "    Return iterator for fasta\n",
    "    \"\"\"\n",
    "    \n",
    "    \"first open the file outside \"\n",
    "    fh = open(fasta_name)\n",
    "    # ditch the boolean (x[0]) and just keep the header or sequence since\n",
    "    # we know they alternate.\n",
    "    faiter = (x[1] for x in groupby(fh, lambda line: line[0] == \">\"))\n",
    "    \n",
    "    for header in faiter:\n",
    "        # drop the \">\"\n",
    "        headerStr = header.__next__()   #[1:].strip()\n",
    "        # join all sequence lines to one.\n",
    "        seq = \"\".join(s for s in faiter.__next__())  # .strip()\n",
    "        yield (headerStr, seq) \n",
    "        \n",
    "def fasta_iter(fasta_name):\n",
    "    \"\"\"\n",
    "    Return iterator for fasta\n",
    "    \"\"\"\n",
    "    \"first open the file outside \"\n",
    "    fh = open(fasta_name)\n",
    "    # ditch the boolean (x[0]) and just keep the header or sequence since\n",
    "    # we know they alternate.\n",
    "    faiter = (x[1] for x in groupby(fh, lambda line: line[0] == \">\"))\n",
    "\n",
    "    for header in faiter:\n",
    "        # drop the \">\"\n",
    "        headerStr = header.__next__()[1:].strip()\n",
    "        # join all sequence lines to one.\n",
    "        seq = \"\".join(s.strip() for s in faiter.__next__())\n",
    "        yield (headerStr, seq) \n",
    "        \n",
    "def clean_seq_line(seq_line, replace_char=r'[^ACGTN\\n\\r]',\n",
    "                   with_char=\"N\"):\n",
    "    \"\"\"Clean seq_line String.\n",
    "    Return updated string\"\"\"\n",
    "    seq_line = seq_line.upper()   # Make Upper Case\n",
    "    seq_line = re.sub(replace_char, with_char, seq_line)\n",
    "    return seq_line\n",
    "\n",
    "def include_samples(sample_seq, min_cov_bs):\n",
    "    \"\"\"Quick check whether to include sample.\n",
    "    min_cov_bs: Minimum Number of covered Bases\n",
    "    return include status, nr covered bases\"\"\"\n",
    "    lgth = len(sample_seq) - sample_seq.count('\\n') - sample_seq.count(\"N\")\n",
    "    include = True\n",
    "    if lgth < min_cov_bs:\n",
    "        include = False\n",
    "    return include, lgth\n",
    "        \n",
    "def split_fasta(fasta_name, path_out=\"./output/singleseq/\", \n",
    "                iids=[], clean=True, \n",
    "                min_cov_bs=20000, output=True):\n",
    "    \"\"\"\n",
    "    Splits up samples from fasta if IID string contains\n",
    "    clean: Whether to clean string (make everything upper-case \n",
    "    and substitute with N)\n",
    "    min_cov_bs: How many bases covered at least once\n",
    "    Return list of samples that have been run\n",
    "    \"\"\"\n",
    "    fiter = fasta_iter_raw(fasta_name)\n",
    "    \n",
    "    iids_org, iids_prcsd = [], []\n",
    "    cvg, icld = [], []\n",
    "    \n",
    "    for sample_line, seq_line in fiter:\n",
    "        sample = sample_line[1:].strip()\n",
    "        if sample in iids:\n",
    "            if output:\n",
    "                print(f\"Matched: {sample}\")\n",
    "            sample1 = re.sub(\"[\\|\\/]\", \".\", sample)\n",
    "            \n",
    "            if clean:\n",
    "                seq_line = clean_seq_line(seq_line)\n",
    "                \n",
    "            include, lgth = include_samples(seq_line, min_cov_bs=min_cov_bs)\n",
    "            \n",
    "            ### Append Statistics\n",
    "            iids_org.append(sample)\n",
    "            iids_prcsd.append(sample1)\n",
    "            cvg.append(lgth)\n",
    "            icld.append(include)\n",
    "            \n",
    "            if include:\n",
    "                fasta_name_out = os.path.join(path_out, sample1 + \".fasta\")\n",
    "                write_fasta(sample_line, seq_line, fasta_name_out, output=output)\n",
    "                \n",
    "    # Make a Summary Dataframe\n",
    "    df = pd.DataFrame({\"iid\":iids_org, \"iid_clean\":iids_prcsd, \n",
    "                        \"cov\":cvg,  \"include\":icld})\n",
    "    return df\n",
    "                  \n",
    "def clean_fasta(fasta_name, fasta_name_out):\n",
    "    \"\"\"\n",
    "    Remove invalid characters in sequences\n",
    "    \"\"\"\n",
    "    \"first open the file outside \"\n",
    "    f = open(fasta_name)\n",
    "    f1 = open(fasta_name_out, \"w\")\n",
    "    n=0\n",
    "    \n",
    "    for line in f:\n",
    "        if line[0]==\">\":\n",
    "            n+=1\n",
    "        else:   # If Genome Data replace i\n",
    "            line = line.replace(\"i\", \"N\")\n",
    "        f1.write(line)  # Write the new line\n",
    "    f.close()\n",
    "    f1.close()\n",
    "    print(f\"Successfully Modified {n} fastas\")\n",
    "    print(f\"Saved to {clean_fasta_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split up the Multi FASTA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 9.58 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "df = split_fasta(fasta_name=\"./data/mar20/gisaid_cov2020_sequences_26.fasta\", \n",
    "            path_out=\"./output/singleseq/\", \n",
    "            iids=iids[:500], clean=True, min_cov_bs=20000, output=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved Summary Dataframe n=500 to ./output/single_seq_split.tsv\n"
     ]
    }
   ],
   "source": [
    "df[\"include\"].value_counts()\n",
    "savepath = \"./output/single_seq_split.tsv\"\n",
    "df.to_csv(savepath, sep=\"\\t\")\n",
    "print(f\"Saved Summary Dataframe n={len(df)} to {savepath}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Dataframe of alredy processed Individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['hCoV-19.USA.WA-S88.2020.EPI_ISL_417141.2020-03-01',\n",
       " 'hCoV-19.USA.WA-S89.2020.EPI_ISL_417142.2020-02-29',\n",
       " 'hCoV-19.USA.WA-S87.2020.EPI_ISL_417140.2020-03-01',\n",
       " 'hCoV-19.USA.WA-S92.2020.EPI_ISL_417145.2020-02-29',\n",
       " 'hCoV-19.USA.WA-S93.2020.EPI_ISL_417146.2020-02-29',\n",
       " 'hCoV-19.USA.WA-S90.2020.EPI_ISL_417143.2020-02-29',\n",
       " 'hCoV-19.USA.WA-S91.2020.EPI_ISL_417144.2020-03-02',\n",
       " 'hCoV-19.USA.WA-S96.2020.EPI_ISL_417149.2020-02-28',\n",
       " 'hCoV-19.USA.WA-S94.2020.EPI_ISL_417147.2020-02-28',\n",
       " 'hCoV-19.USA.WA-S95.2020.EPI_ISL_417148.2020-02-28']"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully Modified 1871 fastas\n",
      "Saved to ./data/mar20/gisaid_cov2020_sequences_26_c.fasta\n",
      "Wall time: 1.81 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "bin_path = \"bins/mafft-win/mafft.bat\"\n",
    "out_path = \"./output/aligned/test_out.fasta\"\n",
    "ref_path = \"./data/reference/wuhan-hu-1.fasta\"\n",
    "\n",
    "fasta_path = \"./data/mar20/gisaid_cov2020_sequences_26.fasta\"\n",
    "clean_fasta_path = \"./data/mar20/gisaid_cov2020_sequences_26_c.fasta\"\n",
    "threads = 4\n",
    "\n",
    "clean_fasta(fasta_path, clean_fasta_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract Single Sequence Fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "split_fasta() got an unexpected keyword argument 'fasta_name_out'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<timed exec>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: split_fasta() got an unexpected keyword argument 'fasta_name_out'"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "split_fasta(fasta_name = \"./data/mar20/gisaid_cov2020_sequences_26.fasta\",\n",
    "            path_out=\"./output/singleseq/\",\n",
    "            fasta_name_out = fasta_name_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_sequence(bin_path, out_path, fasta_name_out, \n",
    "                   ref_path, bat_path=\"./bins/mafft_batch.bat\"):\n",
    "    \"\"\"Produce a batch file and runs bafft\"\"\"\n",
    "    command = f\"{bin_path} --out {out_path} --add {fasta_name_out} --mapout {ref_path}\"\n",
    "    #command=\"echo test\"\n",
    "    print(\"running command:\")\n",
    "    print(command)\n",
    "    stream = os.popen(command)\n",
    "    output = stream.read()\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running command:\n",
      "bins/mafft-win/mafft.bat --out ./output/aligned/test_out_single.fasta --addfull ./output/singleseq/test.fasta --mapout ./data/reference/wuhan-hu-1.fasta\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "align_sequence(bin_path, out_path, fasta_name_out, ref_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align all fasta files\n",
    "Idea: Produce batch script that aligns them one by one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "bin_path = \"bins/mafft-win/mafft.bat\"\n",
    "out_path = \"./output/aligned/test_out.fasta\"\n",
    "ref_path = \"./data/reference/wuhan-hu-1.fasta\"\n",
    "clean_fasta_path = \"./data/mar20/gisaid_cov2020_sequences_26_c.fasta\"\n",
    "thread = 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run this command in (power) shell:\n",
      "bins/mafft-win/mafft.bat --out ./output/aligned/test_out.fasta --add ./data/mar20/gisaid_cov2020_sequences_26_c.fasta --thread 4 --mapout ./data/reference/wuhan-hu-1.fasta\n"
     ]
    }
   ],
   "source": [
    "print(\"Run this command in (power) shell:\")\n",
    "print(f\"{bin_path} --out {out_path} --add {clean_fasta_path} --thread {thread} --mapout {ref_path} --keeplength\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Area 51"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "fiter = fasta_iter(\"./data/mar20/gisaid_cov2020_sequences_26.fasta\")\n",
    "iids = np.array([ff[0] for ff in fiter])\n",
    "fiter = fasta_iter_raw(\"./data/mar20/gisaid_cov2020_sequences_26.fasta\")\n",
    "seqs = np.array([ff[1] for ff in fiter])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample_line, seq_line = next(fiter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "covered = [len(s) - s.count('\\n') - s.count(\"N\") for s in seqs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls = [set(s) for s in seqs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_seq_line(seqs[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick Coverage Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "fiter = fasta_iter_raw(\"./data/mar20/gisaid_cov2020_sequences_26.fasta\")\n",
    "seqs = np.array([ff[1] for ff in fiter])\n",
    "covered = [len(s) - s.count('\\n') - s.count(\"N\") for s in seqs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD7CAYAAACRxdTpAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAUEElEQVR4nO3de4xW+X3f8fcnsMa3tIbsQKfAGlLhtKxVr+Mpdeq0cowdsLsymz9WxWoi1G5FL9SyrV4WailVKiGtL6rSqt2kKE5LlY3JxLG1aOXaJjTbNlKyeHa9ay+7Sxl7L8zCwsSW67ZJ14Z8+8dz8D47PMM8AzNcfvN+Seic831+58z3N8BnzpzzXFJVSJLa8iPXuwFJ0sIz3CWpQYa7JDXIcJekBhnuktQgw12SGjRUuCf5WJLjSZ5M8tkkr02yKsmRJCe75cq+8fuSTCY5kWTb4rUvSRokcz3PPcla4PeBzVX1J0nGgS8Cm4HvVNV9SfYCK6vq3iSbgc8CW4A/D/wu8JaqurCYE5EkvWL5PMa9LskPgNcDp4F9wLu7xw8CDwP3AjuAQ1X1MvBskkl6Qf8Hsx381ltvrQ0bNlxB+5K0dD366KN/VFUjgx6bM9yr6sUknwZeAP4E+EpVfSXJmqo60405k2R1t8ta4A/7DjHV1Wa1YcMGJiYmhpiKJOmiJM/P9tic19y7a+k7gI30LrO8IcnPX26XAbVLrv0k2Z1kIsnE9PT0XG1IkuZhmBuq7wWerarpqvoB8HngrwFnk4wCdMtz3fgpYH3f/uvoXcZ5lao6UFVjVTU2MjLwtwpJ0hUaJtxfAN6Z5PVJAmwFngYOA7u6MbuAB7v1w8DOJCuSbAQ2AccWtm1J0uUMc839kSSfAx4DzgNfAw4AbwTGk9xD7wfA3d34490zap7qxu/xmTKSdG3N+VTIa2FsbKy8oSpJ85Pk0aoaG/SYr1CVpAYZ7pLUIMNdkhpkuEtSgwx3SRrS6LrbSEISRtfddr3buaxh31tGkpa8l148xZvvfQiA5z9x53Xu5vI8c5ekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSg+YM9yQ/keTxvj/fS/LRJKuSHElysluu7NtnX5LJJCeSbFvcKUiSZpoz3KvqRFXdUVV3AO8A/hj4ArAXOFpVm4Cj3TZJNgM7gduB7cD9SZYtUv+SpAHme1lmK/DNqnoe2AEc7OoHgbu69R3Aoap6uaqeBSaBLQvRrCRpOPMN953AZ7v1NVV1BqBbru7qa4FTfftMdTVJ0jUydLgneQ3wQeC35xo6oFYDjrc7yUSSienp6WHbkCQNYT5n7u8HHquqs9322SSjAN3yXFefAtb37bcOOD3zYFV1oKrGqmpsZGRk/p1LkmY1n3D/EK9ckgE4DOzq1ncBD/bVdyZZkWQjsAk4drWNSpKGN9RnqCZ5PfA+4O/3le8DxpPcA7wA3A1QVceTjANPAeeBPVV1YUG7liRd1lDhXlV/DPzYjNq36T17ZtD4/cD+q+5OknRFfIWqJDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGDRXuSd6U5HNJnknydJKfSrIqyZEkJ7vlyr7x+5JMJjmRZNvitS9JGmTYM/d/A3ypqv4i8DbgaWAvcLSqNgFHu22SbAZ2ArcD24H7kyxb6MYlSbObM9yT/BngbwCfAaiq71fVd4EdwMFu2EHgrm59B3Coql6uqmeBSWDLQjcuSZrdMGfuPw5MA/8xydeS/FqSNwBrquoMQLdc3Y1fC5zq23+qq0mSrpFhwn058JPAr1TV24H/S3cJZhYZUKtLBiW7k0wkmZienh6qWUnScIYJ9ylgqqoe6bY/Ry/szyYZBeiW5/rGr+/bfx1weuZBq+pAVY1V1djIyMiV9i9JGmDOcK+ql4BTSX6iK20FngIOA7u62i7gwW79MLAzyYokG4FNwLEF7VqSdFnLhxz3YeCBJK8BvgX8HXo/GMaT3AO8ANwNUFXHk4zT+wFwHthTVRcWvHNJ0qyGCveqehwYG/DQ1lnG7wf2X0VfkqSr4CtUJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1aKhwT/Jckm8keTzJRFdbleRIkpPdcmXf+H1JJpOcSLJtsZqXJA02nzP3n6mqO6rq4mep7gWOVtUm4Gi3TZLNwE7gdmA7cH+SZQvYsyRpDldzWWYHcLBbPwjc1Vc/VFUvV9WzwCSw5Sq+jiRpnoYN9wK+kuTRJLu72pqqOgPQLVd39bXAqb59p7qaJOkaWT7kuHdV1ekkq4EjSZ65zNgMqNUlg3o/JHYD3HbbbUO2IUkaxlBn7lV1ulueA75A7zLL2SSjAN3yXDd8Cljft/s64PSAYx6oqrGqGhsZGbnyGUiSLjFnuCd5Q5IfvbgO/CzwJHAY2NUN2wU82K0fBnYmWZFkI7AJOLbQjUuSZjfMZZk1wBeSXBz/m1X1pSRfBcaT3AO8ANwNUFXHk4wDTwHngT1VdWFRupckDTRnuFfVt4C3Dah/G9g6yz77gf1X3Z0k6Yr4ClVJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0aOtyTLEvytSQPddurkhxJcrJbruwbuy/JZJITSbYtRuOSpNnN58z9I8DTfdt7gaNVtQk42m2TZDOwE7gd2A7cn2TZwrQrSRrGUOGeZB3wN4Ff6yvvAA526weBu/rqh6rq5ap6FpgEtixMu5KkYQx75v7LwD8H/rSvtqaqzgB0y9VdfS1wqm/cVFeTJF0jc4Z7kjuBc1X16JDHzIBaDTju7iQTSSamp6eHPLQkaRjDnLm/C/hgkueAQ8B7kvwGcDbJKEC3PNeNnwLW9+2/Djg986BVdaCqxqpqbGRk5CqmIEmaac5wr6p9VbWuqjbQu1H6X6vq54HDwK5u2C7gwW79MLAzyYokG4FNwLEF71ySNKureZ77fcD7kpwE3tdtU1XHgXHgKeBLwJ6qunC1jUrS9TC67jaSkAy64nzjWj6fwVX1MPBwt/5tYOss4/YD+6+yN0m67l568RRvvvchAJ7/xJ3XuZvh+QpVSWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNmjPck7w2ybEkTyQ5nuSXuvqqJEeSnOyWK/v22ZdkMsmJJNsWcwKSpEsNc+b+MvCeqnobcAewPck7gb3A0araBBzttkmyGdgJ3A5sB+5PsmwxmpckDTZnuFfP/+k2b+n+FLADONjVDwJ3des7gENV9XJVPQtMAlsWtGtJ0mUNdc09ybIkjwPngCNV9QiwpqrOAHTL1d3wtcCpvt2nupok6RoZKtyr6kJV3QGsA7YkeetlhmfQIS4ZlOxOMpFkYnp6erhuJUlDmdezZarqu8DD9K6ln00yCtAtz3XDpoD1fbutA04PONaBqhqrqrGRkZEraF2SNJthni0zkuRN3frrgPcCzwCHgV3dsF3Ag936YWBnkhVJNgKbgGML3bgkaXbLhxgzChzsnvHyI8B4VT2U5A+A8ST3AC8AdwNU1fEk48BTwHlgT1VdWJz2JUmDzBnuVfV14O0D6t8Gts6yz35g/1V3J0m6Ir5CVZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSg4b5gOz1SX4vydNJjif5SFdfleRIkpPdcmXfPvuSTCY5kWTbYk5AknSpYc7czwP/pKr+EvBOYE+SzcBe4GhVbQKOdtt0j+0Ebge2A/d3H64tSbpG5gz3qjpTVY916/8beBpYC+wADnbDDgJ3des7gENV9XJVPQtMAlsWunFJ0uzmdc09yQbg7cAjwJqqOgO9HwDA6m7YWuBU325TXU2SdI0MHe5J3gj8DvDRqvre5YYOqNWA4+1OMpFkYnp6etg2JElDGCrck9xCL9gfqKrPd+WzSUa7x0eBc119Cljft/s64PTMY1bVgaoaq6qxkZGRK+1fkjTAMM+WCfAZ4Omq+td9Dx0GdnXru4AH++o7k6xIshHYBBxbuJYlSXNZPsSYdwG/AHwjyeNd7V8A9wHjSe4BXgDuBqiq40nGgafoPdNmT1VdWPDOJUmzmjPcq+r3GXwdHWDrLPvsB/ZfRV+SpKvgK1QlqUGGuyQ1yHCXpAYZ7pLUIMNdkq7EsltIQhJG1912vbu5xDBPhZQkzXThB7z53ocAeP4Td17nZi7lmbskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGjTMB2T/epJzSZ7sq61KciTJyW65su+xfUkmk5xIsm2xGpckzW6YM/f/BGyfUdsLHK2qTcDRbpskm4GdwO3dPvcnWbZg3UqShjJnuFfVfwe+M6O8AzjYrR8E7uqrH6qql6vqWWAS2LJAvUqShnSl19zXVNUZgG65uquvBU71jZvqapKka2ihb6hmQK0GDkx2J5lIMjE9Pb3AbUjS0nal4X42yShAtzzX1aeA9X3j1gGnBx2gqg5U1VhVjY2MjFxhG5KkQa403A8Du7r1XcCDffWdSVYk2QhsAo5dXYuStLhG1932w89DvVE/E3W+5vwM1SSfBd4N3JpkCviXwH3AeJJ7gBeAuwGq6niSceAp4Dywp6ouLFLvkrQgXnrx1A8/DxVuzM9Ena85w72qPjTLQ1tnGb8f2H81TUmSro6vUJWkBhnuktQgw13SktR/E7VFhrukJeniTdT+G6ktmfOGqiQtOctuuenP6D1zl6SZLvzgpj+rN9wlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJTWt/z1kWvgQjmH59gOSmtb/QRzPf/rnbvq3FRiWZ+6Slo4G3lZgWIa7pJvWUr3kMoxFC/ck25OcSDKZZO9ifR1JS1f/2/a+9NJLPwz65Ste1/R7tQ9jUcI9yTLg3wPvBzYDH0qyeTG+liQBr7rkcuH7/2/JXH6ZzWKduW8BJqvqW1X1feAQsGORvpakGW6UyxWz9TFMf/1j+s/EPSsfzmKF+1rgVN/2VFfTDeBG+Y9/Lc03TBbj+3I1QTfMMftDb5jLFTO3F+r70j/+VX28+EokDNNf/5j+M3HPyoeTqlr4gyZ3A9uq6u91278AbKmqD/eN2Q3s7jbfCjy54I3c2G4F/uh6N3GNOeelwTlfO2+uqpFBDyzW89yngPV92+uA0/0DquoAcAAgyURVjS1SLzck57w0OOel4Uac82JdlvkqsCnJxiSvAXYChxfpa0mSZliUM/eqOp/kHwNfBpYBv15Vxxfja0mSLrVobz9QVV8Evjjk8AOL1ccNzDkvDc55abjh5rwoN1QlSdeXbz8gSQ1asHBPsj7J7yV5OsnxJB/p6quSHElysluu7NtnX/f2BCeSbOurvyPJN7rH/m26VyskWZHkt7r6I0k2LFT/V+Iyc/5UkmeSfD3JF5K8qW+fJufc9/g/TVJJbu2rNTvnJB/u5nU8ySf76k3OOckdSf4wyeNJJpJs6dvnpp1zktcmOZbkiW6+v9TVb978qqoF+QOMAj/Zrf8o8D/pvfXAJ4G9XX0v8IlufTPwBLAC2Ah8E1jWPXYM+CkgwH8B3t/V/xHwq936TuC3Fqr/BZ7zzwLLu/onlsKcu+319G6iPw/c2vqcgZ8BfhdY0T22egnM+St9PX8AeLiFOXe9vbFbvwV4BHgnN3F+LeY360HgfcAJYLTvH8yJbn0fsK9v/Je7b8go8Exf/UPAf+gf060vp/eigVzP/wSD5jyj9nPAA0thzsDngLcBz/FKuDc7Z2AceO+Ax1ue85eBv9XX/2+2Nmfg9cBjwF+9mfNrsd44bAPwdno//dZU1RmAbrm6GzbbWxSs7dZn1l+1T1WdB/4X8GOLMYf5mjHnfn+X3k9vaHjOST4IvFhVT8wY1uycgbcAf737Ffu/Jfkr3bCW5/xR4FNJTgGfphdy0MCckyxL8jhwDjhSVTd1fi14uCd5I/A7wEer6nuXGzqgVpepX26f62q2OSf5OHAeeOBiacDuN/2c6c3x48AvDho6oHbTz7n7e14OrKT36/s/A8a766stz/kfAh+rqvXAx4DPXBw6YPebas5VdaGq7qD3ivotSd56meE3/HwXNNyT3ELvH8IDVfX5rnw2yWj3+Ci9n4ow+1sUTHXrM+uv2ifJcuDPAt9ZyDnM1yxzJsku4E7gb1f3exjtzvkv0Lvu+ESS5+j1/1iSP0e7c4Zen5+vnmPAn9J7j5GW57wLuLj+2/TeARYamTNAVX0XeBjYzs2cXwt4nSrAfwZ+eUb9U7z6hsQnu/XbefUNiW/xyg2Jr9I7G7p4Q+IDXX0Pr74hMX6dr83NNuftwFPAyIx6s3OeMeY5Xrnm3uycgX8A/Ktu/S30fuVO43N+Gnh3t74VeLSFv2dgBHhTt/464H/QOzm7afNrIb85P03vV4yvA493fz5A75rSUeBkt1zVt8/H6d1lPkF3R7mrj9F7l8hvAv+OV15s9Vp6ZwuT9O5I//h1/g8w25wnu//oF2u/2vqcZ4x5ji7cW54z8BrgN7o5PAa8ZwnM+aeBR+kF2yPAO1qYM/CXga91830S+MWuftPml69QlaQG+QpVSWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoP+PxY2AEoewDzPAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "bins = np.arange(0, 31000, 100)\n",
    "plt.figure()\n",
    "ax = plt.gca()\n",
    "ax.hist(covered, bins=bins, ec=\"k\")\n",
    "ax.set_xlim([20000,31000])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "re.sub()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
